components/post_view/post_reaction/post_reaction.tsx

```diff
@@ -48,10 +54,30 @@ export default class PostReaction extends React.PureComponent<Props, State> {
         showEmojiPicker: false,
     };
 
-    handleAddEmoji = (emoji: Emoji): void => {
+    handleEmoji = (emoji: Emoji): void => {
         this.setState({showEmojiPicker: false});
         const emojiName = 'short_name' in emoji ? emoji.short_name : emoji.name;
-        this.props.actions.addReaction(this.props.postId, emojiName);
+        const state = store.getState();
+        const getReactionsForPost = makeGetUniqueReactionsToPost();
+
+        const reactions = getReactionsForPost(state, this.props.postId);
+        let currentUserReacted = false;
+
+        for (const key in reactions) {
+            if ({}.hasOwnProperty.call(reactions, key)) {
+                const value = reactions[key];
+                if (value.user_id === getCurrentUserId(state) && value.emoji_name === emojiName) {
+                    currentUserReacted = true;
+                    break;
+                }
+            }
+        }
+
+        if (currentUserReacted) {
+            this.props.actions.removeReaction(this.props.postId, emojiName);
+        } else {
+            this.props.actions.addReaction(this.props.postId, emojiName);
+        }
```

Comment:
Since we're repeating this process in a few different components, perhaps you could make a `toggleReaction` action that checks if the user reacted with that emoji and calls `addReaction` or `removeReaction` accordingly? That way, we don't need to repeat this logic and each component can just dispatch a `toggleReaction` instead of having to dispatch either an add or remove.

Regarding my previous comment, if `toggleReaction` is a thunk action (ie one that returns a function that looks like `(dispatch, getState) => ...`), then we can also get the state needed to calculate `currentUserReacted` from its `getState` without accessing the global store instance directly.

Reply:
Hello @hmhealey, I am working on this issue with @yoojink0913. We assumed the duplicate code would be an issue as it is bad style and would certainly be in favor of abstracting away the logic as you described. However we are also new to React-redux (I have used React-Javascript before so I am still familiar) and were uncertain with how to abstract this logic without getting all sorts of errors. We attempted to do such in the index.ts files for the components but we ultimately opted to get the toggle behavior working first before figuring out some of the React-redux magic. If you could point me in the right direction, or give me an example of an action like you described that would be much appreciated.

Reply:
Looking at https://github.com/mattermost/mattermost-webapp/pull/10186, it looks like you've got the right idea with using the [redux-thunk action](https://github.com/reduxjs/redux-thunk) to access the store and figure out whether or not it needs to add or remove the reaction.